### ISSUE TITLE
Unable to authorize request

### ISSUE THREAD

**Issue by nathanhamilton** [grokify/ringcentral-sdk-ruby#1]
I will have to research this further myself at a different time, but when I follow the instructions to authorize my application with the password grant method and I call `client.authorize_password('myUsername', 'myExtension', 'myPassword')` it returns this error: 

`NoMethodError: undefined method `+' for nil:NilClass
from /Users/myusernamehere/.rbenv/versions/2.2.3/lib/ruby/2.2.0/net/http.rb:1532:in `addr_port'`

I might have a gem out of date or something similar but any direction you can give me would be helpful. Thanks in advance!

**Comment by nathanhamilton** [comment]
So evidently something is not being passed into Faraday since it uses the `Net::HTTP` with ruby because that line in the ruby source code here https://github.com/ruby/ruby/blob/ruby_2_2/lib/net/http.rb#L1532 
where it is checking to see if it should use and SSL or no.

**Comment by nathanhamilton** [comment]
Ok, so I tried something different. Instead of passing in the url I wanted to point it to I simply initialized the connection like this: 

`client = RingCentralSdk.new(ENV['RING_CENTRAL_API_KEY'], ENV['RING_CENTRAL_API_SECRET'])` 

and then tried to authenticate again as documented: 

`client.authorize(ENV['RING_CENTRAL_PHONE'], ENV['RING_CENTRAL_PASSWORD'])` 

and it returned a different error: 

`OAuth2::Error: invalid_client: Authentication header is missing or empty
{
  "error" : "invalid_client",
  "error_description" : "Authentication header is missing or empty"
}
from /Users/nathanhamilton/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/oauth2-1.2.0/lib/oauth2/client.rb:113:in `request'`

**Comment by grokify** [comment]
Can you post your dependency versions?

I just tested `client.authorize_password()` and the alias `client.authorize()` on Ruby 2.2.3, 2.2.2, and 2.3.0 without issue.

**Comment by nathanhamilton** [comment]
Hey @grokify  Thanks for your help with this.

Can you specify what you are needing when you say dependency versions? I am running this client in a ruby on rails application which is using the following:

ruby 2.2.3p173
Rails 4.2.7
rake, version 11.2.2

**Comment by nathanhamilton** [comment]
OK, so I realized this morning that since it has to do with Ruby's Net HTTP module, the issue probably lay with the URL. After further investigation I realized that I was using undefined environment variables. I have fixed that and am passing the correct information to the SDK. However, now that I have correctly sent the information, it is returning an OAuth2 error.

``` ruby
OAuth2::Error: unauthorized_client: Unauthorized for this grant type
{
  "error" : "unauthorized_client",
  "error_description" : "Unauthorized for this grant type"
}
from /Users/myaccount/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/oauth2-1.2.0/lib/oauth2/client.rb:113:in `request'
```

Doesn't the SDK handle the type of grant being passed for authentication? Let me know if there is any information I can provide that would help.

**Comment by grokify** [comment]
This error typically occurs when the grant type you are using for your app isn't supported for your app's [platform type](https://developer.ringcentral.com/library/tutorials/app-development.html#/#PlatformTypes) in the [Developer Portal](https://developers.ringcentral.com).

The SDK supports two [OAuth 2.0](https://tools.ietf.org/html/rfc6749) grant types:

| OAuth 2.0 Flow | Description |
| --- | --- |
| [Authorization Code Flow](https://developer.ringcentral.com/api-docs/latest/index.html?section=AuthorizationFlows.html#!#AuthorizationCodeFlow) | This flow has a UI similar to Facebook login, etc. It is useful for end-user apps with a web UI. A [Sinatra demo is available using the SDK](https://github.com/grokify/ringcentral-demos-oauth/tree/master/ruby-sinatra). |
| [Password Flow](https://developer.ringcentral.com/api-docs/latest/index.html?section=AuthorizationFlows.html#!#PasswordFlow) (aka ROPC - Resource Owner Password Credentials) | This flow is useful for backend apps (no UI) and test apps. Sample code is available in the SDK's `scripts` directory |

For a Rails app, you can select the "Server/Web" type and should be able to use either flow if you are creating a private app for your organization only, for example:

> Server/Web — web-based apps are apps that communicate with RingCentral APIs from the server, e.g. Node.js, PHP / Laravel, Ruby on Rails, etc.

Public apps to be used across more than one organization need to use Authorization Code flow for security purposes.

**Comment by nathanhamilton** [comment]
Very helpful info. Thank you! So I checked our settings and we do have the platform type set to `Server/Web` in the ring central development app settings and yet it is still unable to authenticate.  I am currently using the password type authentication for our application. Do you have any suggestions?

**Comment by nathanhamilton** [comment]
Hey @grokify, Thanks for your help. I actually did figure out the issue. For my purpose there was going to be no user interface needed to have a user login to authenticate so the appropriate platform type should have been `Server only`. I also realized that I was not including the `+1` characters in the account phone number. Creating a new application with the platform `Server Only` and changing the phone number to be updated allowed me to successfully authenticate and send text messages. Thanks again for your help!
